### ISSUE TITLE
HTML CodeSniffer 'undefined' error

### ISSUE THREAD

**Issue by blobula** [pa11y/pa11y#234]
Hello, 

I recently ran across this tool and I'm trying to get it set up to do some accessibility testing against a site. My issue is when I try to run the following command I'm getting this error.

HTML CodeSniffer: undefined is not an object (evaluating 'f.nodetype') - I'm not understanding what this means. Do I need to set something else up?

I'm running this via the command line prompt in Node.js
```
pa11y --debug --config \npm\node_mo
dules\pa11y\example\before-script\index.js https://mysite/edituseraccount?selectTab=Login > test.txt
```

It first runs the file to log into the site and then should test the edituseraccount page.

This is my before script
```
var test = pa11y({

	// Log what's happening to the console
	log: {
		debug: console.log.bind(console),
		error: console.error.bind(console),
		info: console.log.bind(console)
	},

	// A script to be run on the inital page before Pa11y is run.
	// beforeScript accepts three parameters, the page object, the pa11y options and a callback
	beforeScript: function(page, options, next) {
		
		// Phantom helpers
        page.onConsoleMessage = function(msg) {
            // this allows you to use console.log() inside page.evaluate() to log to your NodeJS console
            console.log(msg);
        };
        page.onError = function(msg, trace) {
          // this will log an error to your NodeJS console if any JS errors occur on the page being tested
          var msgStack = ['ERROR: ' + msg];
          if (trace && trace.length) {
            msgStack.push('TRACE:');
            trace.forEach(function(t) {
              msgStack.push(' -> ' + t.file + ': ' + t.line + (t.function ? ' (in function "' + t.function +'")' : ''));
            });
          }
          console.error(msgStack.join('\n'));
        };

		// An example function that can be used to make sure changes have been confirmed before continuing to run Pa11y
		var waitUntil = function(condition, retries, waitOver) {
			page.evaluate(condition, function(error, result) {
				if (result || retries < 1) {
					// Once the changes have taken place continue with Pa11y testing
					waitOver();
				} else {
					retries -= 1;
					setTimeout(function() {
						waitUntil(condition, retries, waitOver);
					}, 200);
				}
			});
		};

		// The script to manipulate the page must be run with page.evaluate to be run within the context of the page
		page.evaluate(function() {
			var user = document.querySelector('#username');
			var password = document.querySelector('#password');
			var submit = document.querySelector('button[type="submit"]');

			user.value = 'userlogin';
			password.value = 'passwordhere';

			submit.click();

		}, function() {

			// Use the waitUntil function to set the condition, number of retries and the callback
			waitUntil(function() {
				return window.location.href === 'https://mysites.com/edituseraccount?selectTab=Login';
			}, 20, next);
		});
	}

});
```

**Comment by hollsk** [comment]
Hi @blobula - thanks for using Pa11y!

It looks like you might have run into the same problem as in #165 - maybe try the resolution in there as a first step?

**Comment by blobula** [comment]
Hello, I actually looked at that issue and tried it, however I was unsuccessful in getting it to work for me.

I'm running this on Windows 7.

I noticed that if I run this command I don't have that error.
`pa11y --debug https://mysite/edituseraccount?selectTab=Login > test.txt`

When I add the config file that's when it happens.

**Comment by hollsk** [comment]
Oh, hrm. That is interesting. What's your full error output, please?

**Comment by blobula** [comment]
This is my output. It appears everything goes ok, however there is the CodeSniffer error on the command prompt window, and it's actually never logging into the site and testing the page specified. It just tests the login page. Sorry if I'm missing something. I'm completely new to using Pa11y.

`Welcome to Pa11y

PhantomJS browser created
 > PhantomJS browser created
PhantomJS page created
 > Debug: PhantomJS page created
Opening "https://mysite/login.geo" in PhantomJS
 > Debug: Opening "https://mysite/edituseraccount.myprofile.geo?selectTab=Login" in PhantomJS
Opened "https://mysite/login.geo" in PhantomJS
Testing the page "https://mysite/login.geo"
Running beforeScript
 > Debug: Opened "https://mysite/edituseraccount.myprofile.geo?selectTab=Login" in PhantomJS
 > Testing the page "https://mysite/edituseraccount.myprofile.geo?selectTab=Login"
 > Debug: Injecting HTML CodeSniffer
 > Debug: Injecting Pa11y
 > Debug: Running Pa11y on the page
 > Debug: Document title: "Login"
 > Debug: Test function ran for "https://mysite/edituseraccount.myprofile.geo?selectTab=Login"
 > Debug: PhantomJS page closed
 > Debug: PhantomJS browser exited

Results for https://mysite/edituseraccount.myprofile.geo?selectTab=Login:


5 Errors
5 Warnings
26 Notices

Injecting HTML CodeSniffer
Injecting Pa11y
Running Pa11y on the page
PhantomJS page closed
PhantomJS browser exited
`

This is my entire script file
`// An example of injecting a script before Pa11y runs.
// This example logs in to a fictional site then waits
// until the account page has loaded before running Pa11y
'use strict';


var pa11y = require('../..');

// Create a test instance with some default options
var test = pa11y({

	// Log what's happening to the console
	log: {
		debug: console.log.bind(console),
		error: console.error.bind(console),
		info: console.log.bind(console)
	},

	// A script to be run on the inital page before Pa11y is run.
	// beforeScript accepts three parameters, the page object, the pa11y options and a callback
	beforeScript: function(page, options, next) {
		
		// Phantom helpers
        page.onConsoleMessage = function(msg) {
            // this allows you to use console.log() inside page.evaluate() to log to your NodeJS console
            console.log(msg);
        };
        page.onError = function(msg, trace) {
          // this will log an error to your NodeJS console if any JS errors occur on the page being tested
          var msgStack = ['ERROR: ' + msg];
          if (trace && trace.length) {
            msgStack.push('TRACE:');
            trace.forEach(function(t) {
              msgStack.push(' -> ' + t.file + ': ' + t.line + (t.function ? ' (in function "' + t.function +'")' : ''));
            });
          }
          console.error(msgStack.join('\n'));
        };

		// An example function that can be used to make sure changes have been confirmed before continuing to run Pa11y
		var waitUntil = function(condition, retries, waitOver) {
			page.evaluate(condition, function(error, result) {
				if (result || retries < 1) {
					// Once the changes have taken place continue with Pa11y testing
					waitOver();
				} else {
					retries -= 1;
					setTimeout(function() {
						waitUntil(condition, retries, waitOver);
					}, 200);
				}
			});
		};

		// The script to manipulate the page must be run with page.evaluate to be run within the context of the page
		page.evaluate(function() {
			var user = document.querySelector('#username');
			var password = document.querySelector('#password');
			var submit = document.querySelector('button[type="submit"]');

			user.value = 'valuehere';
			password.value = 'valuehere';

			submit.click();

		}, function() {

			// Use the waitUntil function to set the condition, number of retries and the callback
			waitUntil(function() {
				return window.location.href === 'https://mysite/edituseraccount.myprofile.geo?selectTab=Login';
			}, 20, next);
		});
	}

});

// Start on the http://example.com/login where the beforeScript will be injected
test.run('https://mysite/login.geo', function(error, result) {
	if (error) {
		return console.error(error.message);
	}
	console.log(result);
});`

**Comment by hollsk** [comment]
Oh weird. Your script looks fine, and Pa11y does seem to _think_ it's testing the right page. 

What happens if you use the actions interface instead of beforeScript? e.g.:

```
'use strict';

var pa11y = require('../..);

// Create a test instance with some default options
var test = pa11y({

  // Log what's happening to the console
  log: {
    debug: console.log.bind(console),
    error: console.error.bind(console),
    info: console.log.bind(console)
  },

  actions: [
    'set field #username to exampleUser',
    'set field #password to password1234',
    'click element button[type="submit"]',
    'wait for url to be https://mysite/edituseraccount.myprofile.geo?selectTab=Login'
  ]

});

test.run('https://mysite/login.geo', function(error, result) {
  if (error) {
    return console.error(error.message);
  }
  console.log(result);
});
```

**Comment by blobula** [comment]
When I use the actions interface I no longer get the CodeSniffer undefined error, however per the log it's still not logging into the site and advancing to the designated page that i set for it to test. It just tests the login page and stops.  This is my output showing me it's testing the login page. (I removed some results)

```
Welcome to Pa11y

 > PhantomJS browser created
 > Debug: PhantomJS page created
 > Debug: Opening "https://mysite/edituseraccount.myprofile.geo?selectTab=Login" in PhantomJS
 > Debug: Opened "https://mysite/edituseraccount.myprofile.geo?selectTab=Login" in PhantomJS
 > Testing the page "https://mysite/edituseraccount.myprofile.geo?selectTab=Login"
 > Debug: Injecting HTML CodeSniffer
 > Debug: Injecting Pa11y
 > Debug: Running Pa11y on the page
 > Debug: Document title: "Maestro 5.0: Login"
 > Debug: Test function ran for "https://mysite/edituseraccount.myprofile.geo?selectTab=Login"
 > Debug: PhantomJS page closed
 > Debug: PhantomJS browser exited

Results for https://mysite/edituseraccount.myprofile.geo?selectTab=Login:

 • Notice: Check that the title element describes the document.
   ├── WCAG2AA.Principle2.Guideline2_4.2_4_2.H25.2
   ├── html > head > title
   └── <title> Login </title>

  • Warning: The heading structure is not logically nested. This h4 element should be an h3 to be properly nested.
   ├── WCAG2AA.Principle1.Guideline1_3.1_3_1_A.G141
   ├── #portlet-application > div:nth-child(2) > div > div:nth-child(6) > div:nth-child(1) > div > div > h4
   └── <h4>Login Form:</h4>

5 Errors
5 Warnings
26 Notices
```

**Comment by hollsk** [comment]
Oh, wait - I think I know what's happening :D 

Are you still running `pa11y --debug https://mysite/edituseraccount?selectTab=Login > test.txt`? 

If you're doing that then you're just running the installed command line tool on a single URL, rather than using the javascript interface you've created. You'll want to run `node myjavascriptfile.js` to use the JS interface.

**Comment by blobula** [comment]
Yep, that was the issue! 

Running this works.
`node \npm\node_modules\pa11y\example\before-script\index.js > test.txt`

Thank you!

**Comment by hollsk** [comment]
Yay! Glad it's working now 🎉 

If you have any more problems, let us know in a new issue, happy pa11y-ing 😄

**Comment by blobula** [comment]
Just a couple follow up questions.

Do I need to still pass the page in that I'm wanting to test like this?
`node index.js https://pagetotest.com/edituseraccount.myprofile.geo?selectTab=Login > output.txt`

Also when I output to a text file it seems to stop listing items and just has this at the bottom of the text file.
```
    typeCode: 3 },
  ... 50 more items ]
```
Is it possible to output results via the reporter from within the Javascript file?

**Comment by hollsk** [comment]
Sure - you don't need to pass any of that in because the JS interface is handling everything. Just `node yourfile.js` is enough :-)

You'll need to [use a reporter](https://github.com/pa11y/pa11y#javascript-interface) from inside your javascript file to write out the results. We could do with an example of this in the documentation, but it goes something like this...

require the reporter you want and the filesystem node package at the top of your file:
`var reporter = require('\npm\node_modules\pa11y\reporter\csv');
var fs = require('fs');`

and then inside your `test.run` block do the processing and writing:
```
test.run('https://pagetotest.com/edituseraccount.myprofile.geo?selectTab=Login', function(error, result) {
  if (error) {
    return console.error(error.message);
  }
  var processResults = reporter.process(result);
  fs.writeFile("\path\to\test.txt", processResults, function(err) {
      if(err) {
          return console.log(err);
      }
      console.log("The file was saved!");
  }); 
});
```

Hope that makes sense!
